### PR TITLE
Add validation for zones

### LINKED ISSUE #3952: umbrella issue: deprecating zones in Cluster API
**What would you like to be added**:
This issue tracks subsequent tasks of [deprecating zone with zones](https://github.com/karmada-io/karmada/pull/3933).

- [x] Add validation for `zones`, possibly at [here](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/apis/cluster/validation/validation.go#L152-L164). (@zishen, #3975)
In addition, we should ensure that `zone` and `zones` will not co-exist.
- [x] Transform potential `zone` to `zones`, possibly at [here](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/registry/cluster/strategy.go#L129-L132). (@zishen, #3983)
- [x] Deprecate the [--cluster-zone flag](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/karmadactl/join/join.go#L131) of `join` and introduce a new flag `--cluster-zones`. (@whitewindmills, #3995)
- [x] Adopt `zones` in scheduler spread by constraints, possibly at [here](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/scheduler/core/spreadconstraint/group_clusters.go#L133-L161). (@zishen, #4002)
- [x] Adopt `zones` in scheduler filter, possibly at [here](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/scheduler/framework/plugins/spreadconstraint/spread_constraint.go#L44-L46). (@whitewindmills, #3996)
- [x] Adopt `zones` in resource selector, possibly at [here](https://github.com/karmada-io/karmada/blob/276d6a8ad1f156546a3208aefb26ceb3252c3a7f/pkg/util/selector.go#L113-L116)  (@whitewindmills, #4010)

**Why is this needed**:

### PR DESCRIPTION
**Pull request by zishen**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

 Add validation for zones

**Which issue(s) this PR fixes**:
part of #3952

**Special notes for your reviewer**:

@XiShanYongYe-Chang @whitewindmills 

**Does this PR introduce a user-facing change?**:

none

### CODE REVIEW

**Comment by whitewindmills** [comment]
please don't write like this, it will cause this issue(https://github.com/karmada-io/karmada/issues/3952) to be closed following this PR.
```
Which issue(s) this PR fixes:
Fixes #3952
```

**Comment by zishen** [comment]
> please don't write like this, it will cause this issue(#3952) to be closed following this PR.
> 
> ```
> Which issue(s) this PR fixes:
> Fixes #3952
> ```

Ok, got it.

File: pkg/apis/cluster/validation/validation.go [hunk 1]
```diff
@@ -157,9 +181,11 @@ func validateClusterSpecForFieldSelector(spec *api.ClusterSpec, fldPath *field.P
 	if errs := kubevalidation.IsValidLabelValue(spec.Region); len(errs) != 0 {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("region"), spec.Region, strings.Join(errs, "; ")))
 	}
-	if errs := kubevalidation.IsValidLabelValue(spec.Zone); len(errs) != 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("zone"), spec.Zone, strings.Join(errs, "; ")))
+
+	if errs := validateClusterSpecZoneLabel(spec, fldPath); len(errs) != 0 {
+		allErrs = append(allErrs, errs...)
 	}
```
Comment (RainbowMango) [inline pkg/apis/cluster/validation/validation.go:187]: Given the `.spec.zone` will be removed eventually, so my suggestion is not couple them together, we could do it like:
```golang
	if spec.Zone != "" && len(spec.Zones) > 0 {
		// return errors
	}
	
	if errs := kubevalidation.IsValidLabelValue(spec.Zone); len(errs) != 0 {
		allErrs = append(allErrs, field.Invalid(fldPath.Child("zone"), spec.Zone, strings.Join(errs, "; ")))
	}
	
	for i := range spec.Zones {
		// validate each zones.
	}
```
That, when we remove the .spec.zone, we can easily remove the relevant code.

File: pkg/apis/cluster/validation/validation.go [hunk 2]
```diff
@@ -148,6 +148,30 @@ func ValidateClusterProxyURL(fldPath *field.Path, proxyURL string) field.ErrorLi
...
 
+func validateClusterSpecZoneLabel(spec *api.ClusterSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// Zone and Zones cannot co-exist.
+	// see https://github.com/karmada-io/karmada/issues/3952
+	if spec.Zone != "" && len(spec.Zones) != 0 {
+		return append(allErrs, field.TypeInvalid(fldPath.Child("spec"), spec, "Zone and Zones cannot co-exist"))
+	}
+
+	if spec.Zone != "" {
```
Comment (whitewindmills) [inline pkg/apis/cluster/validation/validation.go:159]: we don't need to validate `.spec.zone` because it will be filled to `.spec.zones` and remain empty.

Comment (RainbowMango) [reply pkg/apis/cluster/validation/validation.go:159]: It doesn't harm though. Both are okay to me.

File: pkg/apis/cluster/validation/validation.go (hunk 1 above)
Comment (zishen) [reply pkg/apis/cluster/validation/validation.go]: good suggestion.

File: pkg/apis/cluster/validation/validation.go (hunk 2 above)
Comment (zishen) [reply pkg/apis/cluster/validation/validation.go:167]: OK,I'll keep it.
And here will be changed later when the zone field is removed.

**Review by RainbowMango — APPROVED** [review]
/approve

@zishen Looks good to me, have you tested it?

**Review by whitewindmills — COMMENTED** [review]
LGTM

**Comment by zishen** [comment]
> /approve
> 
> @zishen Looks good to me, have you tested it?

yes, I had edit cluster, and add zone,zones. All met expectations.

image

**Review by RainbowMango — COMMENTED** [review]
/lgtm

**Pull request merged**